Signal checkpoint patch

pbehne opened the pull request:
This PR patches the issue where a checkpoint can write an unconverged solution depending on when the USR1 signal is received, thereby affecting the accuracy of recovered simulations. The fix 1) adds logic that ensures a checkpoint is not output unless `_current_execute_flag` is contained in `_execute_on`, and 2) enforces that Checkpoint’s `execute_on` parameter may only be set to `TIMESTEP_END` so that only converged solutions are output. Exodiff tests are added to ensure that simulations recovered from signal-based checkpoints give the same solutions as the uninterrupted simulations. The `runCommand` methods of the `Tester` and derived `SignalTester` TestHarness objects are also refactored to eliminate code duplication. Closes #25755.

pbehne committed:
Fix for signal checkpoint writing unconverged soln

This commit patches the issue (#25755) where a checkpoint can write an
unconverged solution depending on when the USR1 signal is received, thereby
affecting the accuracy of recovered simulations. The fix 1) adds logic that
ensures a checkpoint is not output unless _current_execute_flag is contained
in _execute_on, and 2) enforces that Checkpoint’s ‘execute_on’ parameter may
only be set to ‘TIMESTEP_END’ so that only converged solutions are output.
Exodiff tests are added to ensure that simulations recovered from signal-based
checkpoints result in the same solutions as the uninterrupted simulations.

pbehne committed:
typo.

pbehne committed:
Refactored Tester.runCommand and SignalTester.runCommand to eliminate deuplicate code from SignalTester.

pbehne committed:
Removed unused parameters from Tester.runCommand.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1830933/) on 4c9988d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25773/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25773/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bd0458dfc56bc8befa57093b513843437f6b2b0a`

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/1831037/) on 51b304a : invalidated by @pbehne

pbehne commented:
`TestHarness/tests.working_directory` is currently failing everywhere. Is this because of me? @loganharbour

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1831050/) on 51b304a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25773/site)

This comment will be updated on new commits.

loganharbour commented:
> `TestHarness/tests.working_directory` is currently failing everywhere. Is this because of me? @loganharbour

Yeah, it is

lindsayad commented:
unsubscribing while the battle with civet is ongoing

pbehne committed:
Added shell parameters for civet testing.

pbehne commented on python/TestHarness/testers/SignalTester.py:
```diff
@@ -55,57 +55,17 @@ def send_signal(self,pid):
-            timer.start()
-            self.setStatus(self.fail, 'WORKING DIRECTORY NOT FOUND')
-            timer.stop()
+        exit_code = super().spawnSubprocessFromOptions(timer, options, shell_linux=False)
```
Setting `shell_linux=False` is crucial here. If the default value of True is not overridden, then the `test_signal`  test crashes in the `conda (Ubuntu)` check with exit code -10. It seems like sending the signal causes a crash. I cannot replicate this on rod, and do not know why it happens.

loganharbour replied:
Should be able to replicate it now; would be better to know what's actually going on

pbehne replied:
I found the issue. When we set `shell=True`, the test file is run using /bin/sh. When I run the test on my local machine and search for the job, I see this:

`(moose) INL654724:signal_handler behnpa$ ps aux | grep moose_test-opt
behnpa            2887  96.0  1.7 409778640 567376 s000  R     2:42PM   0:09.44 /Users/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3`

In this case, the `SignalTester` is sending the USR1 signal to pid 2887 to write the checkpoint. When I search for the job in the Conda (Ubuntu) environment, I get the following.

`(moose) [ubuntu:20.04-0][~/projects/moose/test/tests/misc/signal_handler]> ps aux | grep moose_test-opt
behnpa    792960  100  0.0 812824 329740 pts/2   Rl   14:35   1:02 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3
behnpa    793006  0.0  0.0   2628   616 pts/2    S    14:36   0:00 /bin/sh -c /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3
behnpa    793007  106  0.0 802756 320120 pts/2   Rl   14:36   0:05 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3
`

So there is a pid for both `/bin/sh` and for `moose_test-opt` itself. When I send a signal to the `sh` pid, the job crashes without writing a checkpoint. When I send a signal to the `moose_test-opt` pid, the checkpoint is written and the job does not crash. `subprocess.Popen` must be setting the pid to the `sh` job, not the `moose_test-opt` job. I'll dig around a bit more to see if there's anything we can do about it instead of setting `shell` to False, but I am fairly satisfied by this explanation.

pbehne replied:
According to [this](https://www.gnu.org/software/libc/manual/html_node/Miscellaneous-Signals.html) documentation, the default action of SIGUSR1 is to terminate the process if signal handling is not set up. I'm guessing that sh does not have signal handling set up, so that is why it crashes, even when I try to propagate the signal to the child process (moose_test-opt).

pbehne replied:
I played around with using signals to trigger a checkpoint other than `SIGUSR1` and passing them to all child processes. I did this because some signals claimed that the default action is to ignore. I has some success with `SIGCHLD` and `SIGPIPE`, but I couldn't get the parallel tests to pass using these. I believe the reason for this is that mpiexec has defined actions for the non-user signals. The processes for the parallel test are as follows:

`(moose) [ubuntu:20.04-0][~/projects/moose/test/tests/misc/signal_handler]> ps aux | grep moose_test-opt
behnpa    806745  0.0  0.0   2628   552 pts/2    T    17:16   0:00 /bin/sh -c mpiexec -n 3 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806746  0.0  0.0   5988  3500 pts/2    S    17:16   0:00 mpiexec -n 3 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806748  102  0.0 731272 240800 ?       Rsl  17:16   0:09 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806749  102  0.0 347980 222320 ?       Rs   17:16   0:09 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806750  102  0.0 346252 219852 ?       Rs   17:16   0:09 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3`

The `mpiexec`s add an additional layer of complexity. Unless someone has an idea, I think the best course of action is to leave `shell=False` for signal test cases.

permcody replied:
Don't overthink it. This is really only useful on the cluster and users may or may not signal the launcher or one of the child processes. Ideally, if we send a signal to any process it'll checkpoint on all ranks and keep going (or stopping cleanly). So I think we should still respond to the signals that don't propagate. We can handle synchronizing all the ranks on our own so no need to rely on something that may or may not work all the time.

dschwen replied:
You could write a file `inputname_out.pid` with the MOOSE root process PID. That's not an uncommon pattern and would make finding the PID really easy for users.

pbehne replied:
So you would propose grabbing the "correct" PID with `ps`, writing it to file, and then grabbing it from file to send the signal? I wonder if it might be even better if we send the `ps` command and process the results in python, storing the correct PID? Of course, we would have to decide if this effort is more worth it (and robust) than just turning the shell parameter off like we are now.

dschwen replied:
No that makes no sense. The MOOSE process knows it's own PID (https://man7.org/linux/man-pages/man2/getpid.2.html) and should write it itself. That's what daemon processes in UNIX write to `/var/run` for example

dschwen replied:
Example:
```
$ l /var/run/*.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/acpid.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/apcupsd.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/crond.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/gdm3.pid
-rw-r--r-- 1 root root 4 Oct 23 15:38 /var/run/libvirtd.pid
-rw-r--r-- 1 root root 4 Oct 23 15:38 /var/run/ntpd.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/sshd.pid
-rw------- 1 root root 5 Oct 23 15:38 /var/run/sssd.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-bridge-0.pid
-rw-r----- 1 root root 5 Oct 23 15:38 /var/run/vmnet-dhcpd-vmnet1.pid
-rw-r----- 1 root root 5 Oct 23 15:38 /var/run/vmnet-dhcpd-vmnet8.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-natd-8.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-netifup-vmnet1.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-netifup-vmnet8.pid
```

pbehne replied:
@permcody, what do you think of this? Could be useful to have the PID sitting in a file while moose runs. I imagine that we would add a task and action (or append to ones that make sense) early on in the setup process that write the pid to file and then delete the file as MOOSE is terminating.

Question: would we want to do this now, or later, when we redesign the output system?

pbehne replied:
This is outdated, since we dropped the shell parameter for most tests.

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -56,6 +61,16 @@ Checkpoint::Checkpoint(const InputParameters & parameters)
     _suffix(getParam<std::string>("suffix"))
 {
+  // Prevent the checkpoint from executing at any time other than TIMESTEP_END
+  ExecFlagEnum execute_on = getParam<ExecFlagEnum>("execute_on");
```
```suggestion
  const auto & execute_on = getParam<ExecFlagEnum>("execute_on");
```

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -47,6 +47,11 @@ Checkpoint::validParams()
 
+  // Since it only makes sense to write checkpoints at the end of time steps,
+  // change the default value of execute_on to TIMESTEP_END
+  ExecFlagEnum & exec_enum = params.set<ExecFlagEnum>("execute_on", true);
+  exec_enum = {EXEC_TIMESTEP_END};
```
```suggestion
  ExecFlagEnum & exec_enum = params.set<ExecFlagEnum>("execute_on", true);
  exec_enum = {EXEC_TIMESTEP_END};
  params.suppressParameter<ExecFlagEnum>("execute_on");
```

Makes it no longer settable by a user

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -56,6 +61,16 @@ Checkpoint::Checkpoint(const InputParameters & parameters)
+  // Prevent the checkpoint from executing at any time other than TIMESTEP_END
+  ExecFlagEnum execute_on = getParam<ExecFlagEnum>("execute_on");
+  if (!(execute_on.size() == 1 && execute_on.contains(EXEC_TIMESTEP_END)))
+  {
+    paramError("execute_on",
+               "The checkpoint system may only be used with execute_on = ",
+               "TIMESTEP_END, not '",
+               execute_on,
+               "'.");
+  }
```
```suggestion
    paramError("execute_on",
               "The checkpoint system may only be used with execute_on = ",
               "TIMESTEP_END, not '",
               execute_on,
               "'.");
```

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7878,6 +7878,10 @@ FEProblemBase::addOutput(const std::string & object_type,
     if (_app.getParam<bool>("show_input") && parameters.get<bool>("output_screen"))
       parameters.set<ExecFlagEnum>("execute_input_on") = EXEC_INITIAL;
   }
+  // Need this because Checkpoint::validParams changes the default value of
+  // execute_on
+  else if (object_type == "Checkpoint")
+    exclude.push_back("execute_on");
```
This shouldn't be needed anymore as we don't apply private parameters

pbehne replied:
Removing this triggers the ParamError, so it seems like that call to applyParameters in FEProblemBase is still overwriting the desired value of `execute_on`.

pbehne replied:
It was decided not to make the param private because we wanted the option to specify INITIAL and/or TIMESTEP_END.

pbehne commented:
The `nan_interface.signaling_nan_opt` test in Modules recover is failing in CIVET, but working fine on my machine. I'll go into the container and see what I can find.

pbehne committed:
Allow execute on INITIAL, added test, fixed test

After discussion with other developers, we allow INITIAL as a valid execute_on
option. The default value is still set to TIMESTEP_END. All other value are
disallowed and will raise a ParamError.

A RunException test called 'execute_on_param_error' was added to test for the
above mentioned ParamError.

Finally, the 'signaling_nan_opt' test in fluid_properties was modified to
not be run in --recover mode because part 1 of the test crashes immediately
without writing a checkpoint (does not perform a timestep). Thus, the --recover
simulation crashes because it can't find a checkpoint file. Hence,
'recover = False' was added to the test.

pbehne commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -62,12 +61,31 @@ Checkpoint::Checkpoint(const InputParameters & parameters)
   const auto & execute_on = getParam<ExecFlagEnum>("execute_on");
-  if (!(execute_on.size() == 1 && execute_on.contains(EXEC_TIMESTEP_END)))
+
+  // Create a vector containing all valid values of execute_on
+  std::vector<ExecFlagEnum> valid_execute_on_values(3);
+  {
+    ExecFlagEnum valid_execute_on_value = execute_on;
+    valid_execute_on_value.clear();
+    valid_execute_on_value += EXEC_INITIAL;
+    valid_execute_on_values[0] = valid_execute_on_value;
+    valid_execute_on_value += EXEC_TIMESTEP_END;
+    valid_execute_on_values[1] = valid_execute_on_value;
+    valid_execute_on_value.clear();
+    valid_execute_on_value += EXEC_TIMESTEP_END;
+    valid_execute_on_values[2] = valid_execute_on_value;
+  }
```
Not a huge fan of the construction of this vector. Would it make sense to implement a constructor for ExecFlagEnum (or one of its parents) so we could do something like
`ExecFlagEnum valid_execute_on_value = {EXEC_INITIAL}`?

Also, since `valid_execute_on_values` should be const after it is constructed, and will be the same for any Checkpoint object, should we find a way to make it a static const attribute?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1845306/) on efcf843 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2dcda</th>
      <th colspan=4>#25773 efcf84</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.64%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92364</td>
      <td>92386</td>
      <td>+22</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15498</td>
      <td>15497</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25773/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25773/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25773/coverage/xfem)

</details>



This comment will be updated on new commits.

pbehne commented:
@loganharbour, is the documentation fail my fault? I know you said the Internal app tests is not. I already have a fix for the Griffin issue in controlled tests.

loganharbour commented:
> @loganharbour, is the documentation fail my fault? I know you said the Internal app tests is not. I already have a fix for the Griffin issue in controlled tests.

See

```
ERRORS:
/data/civet1/build/moose/test/tests/outputs/checkpoint/tests:execute_on_param_error:177
Extra 'requirement' supplied
Control this message using: 'log_extra_requirement' 
```

pbehne committed:
Removed extra requirement from test.

pbehne commented:
Are we having CIVET issues right now? most tests were passing, I made a one line change that should not affect much, and now `civet_results` is failing in a bunch of tests.

loganharbour commented:
> Are we having CIVET issues right now? most tests were passing, I made a one line change that should not affect much, and now `civet_results` is failing in a bunch of tests.

I don't think you develop enough with moose to see that one :) that'll fail when a connection to civet.inl.gov times out. not uncommon

pbehne commented:
@lindsayad @roystgnr @joshuahansel @GiudGiud @loganharbour @dschwen: I have worked all the kinks out of the moose-based apps, so those will all pass once those separate PRs are merged. I am ready for reviews.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1855110/) on e20b0f7 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

GiudGiud commented:
what's going on with banning checkpoint on FINAL or timestep_begin ? do you mind tagging me where this is discussed. thanks!

permcody commented:
We just want to make sure that checkpoint outputs are only output where
they make sense or are valid. I think FINAL makes sense, sort of, so we
probably won't restrict that. Even though timestep_begin "should" be valid,
is there a good reason to output there instead of at the end of a timestep?

On Thu, Nov 2, 2023 at 1:28 PM Guillaume Giudicelli <
***@***.***> wrote:

> what's going on with banning checkpoint on FINAL or timestep_begin ? do
> you mind tagging me where this is discussed. thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/25773#issuecomment-1791422642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOIBNEKRJDU44F5L3V6LYCPX4FAVCNFSM6AAAAAA6GHSKLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJRGQZDENRUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

GiudGiud commented:
Depends when the checkpoint happens. If it s right at the very end of a timestep or right at the very beginning on the next one, then there s not much need

pbehne committed:
Testing with shell=False.

This commit tests all tests without launching them in a separate shell.
A custom string splitting utility function was implemented to correctly split
the commands into lists to be passed to subprocess.Popen. If we procees with
this approach, this utility will need to be cleaned up for conciseness and
clarity. Some of the tests files were changes to reflect the fact that the
tests are no longer run in a separate shell by default.

pbehne committed:
Fixed modules tests to work with shell=False.

pbehne committed:
Now dropping shell and using shlex.split for command splitting.

pbehne committed:
Undoes last 2 commits.

pbehne committed:
Drop shell support in TestHarness, updated tests

The shell parameter in calls to subprocess.Popen has been set to false with no
option to change it. Some tests in the tensor mechanics module were updated
to be compatible with this change.

pbehne committed:
Added 'use_shell' param to Tester

This parameter controls whether subprocess.Popen launches each test using the
shell. The default value is False. For tests where the shell is needed, the
shell may be turned on for each specific test.

pbehne committed:
Undid formatting screw-up.

pbehne commented:
> We just want to make sure that checkpoint outputs are only output where they make sense or are valid. I think FINAL makes sense, sort of, so we probably won't restrict that. Even though timestep_begin "should" be valid, is there a good reason to output there instead of at the end of a timestep?
> […](#)
> On Thu, Nov 2, 2023 at 1:28 PM Guillaume Giudicelli < ***@***.***> wrote: what's going on with banning checkpoint on FINAL or timestep_begin ? do you mind tagging me where this is discussed. thanks! — Reply to this email directly, view it on GitHub <[#25773 (comment)](https://github.com/idaholab/moose/pull/25773#issuecomment-1791422642)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAXFOIBNEKRJDU44F5L3V6LYCPX4FAVCNFSM6AAAAAA6GHSKLWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJRGQZDENRUGI> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Should I go ahead and allow execute on FINAL?

pbehne commented:
@permcody, see my latest changes where I added `use_shell` as a `Tester` parameter with default value False and only enable it for tests that need it.

permcody reviewed (approved):
This looks a lot better to me. I like this solution and think this is good enough to merge even if it's not perfect. It's in the right direction.

pbehne commented:
Once @loganharbour reviews again, I will fix all the moose-based apps that are failing, then this can be merged.

dschwen commented:
Where do you start off when a checkpoint is loaded? If you load a checkpoint that was dumped on timestep begin and load it with no special treatment like a checkpoint that was written on timestep end, wouldn't you end up skipping an entire step?

pbehne commented:
> Where do you start off when a checkpoint is loaded? If you load a checkpoint that was dumped on timestep begin and load it with no special treatment like a checkpoint that was written on timestep end, wouldn't you end up skipping an entire step?

I'm not sure, but I think so. What prompted this PR is that when a USR1 signal is sent to the program, it could dump on LINEAR or NONLINEAR during a timestep. When using that checkpoint to recover, it would start from the next timestep without finishing converging the previous.

dschwen commented:
> When using that checkpoint to recover, it would start from the next timestep without finishing converging the previous.

Ok, I might not have fully caught up, but could we just fail the timestep as recovered and redo it (like when we're cutting the dt). We'd just have to know when that checkpoint was taken.

pbehne commented:
> Ok, I might not have fully caught up, but could we just fail the timestep as recovered and redo it (like when we're cutting the dt). We'd just have to know when that checkpoint was taken.

We have talked about options like this before, i.e., when the signal is received, failing the timestep and writing the checkpoint using the solution at the beginning of the step. This might be more involved in that we may not be currently saving all the necessary states to safety do this and would likely be better for a separate PR. For now, we are just allowing the timestep to finish before writing the checkpoint. This, when combined with an automatic wall time based checkpoint (which I will work on next), should cover most of our bases.

pbehne committed:
Now allow execute on FINAL.

pbehne commented:
@loganharbour App PRs:

Griffin: https://github.inl.gov/ncrc/griffin/pull/1676
Bison: https://github.inl.gov/ncrc/bison/pull/5758
Magpie: https://github.com/idaholab/magpie/pull/494
SAM (already merged): https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/855

This VTB PR also may be relevant, but I cannot find where VTB is tested anymore: https://github.com/idaholab/virtual_test_bed/pull/310

GiudGiud commented:
I don't have time to review right now but looking at the vtb patch, I don't understand why final is no longer allowable. It s the second most useful flag for restart imo

pbehne commented:
> I don't have time to review right now but looking at the vtb patch, I don't understand why final is no longer allowable. It s the second most useful flag for restart imo

@GiudGiud, thanks for bringing this up and jogging my memory. Per the latest commit in this PR, FINAL is allowed. I noticed that I couldn't find where the VTB test was failing anymore, and this is the reason. I'll go ahead and close the VTB PR since it is no longer needed.

GiudGiud commented:
Works for me then!

loganharbour commented:
@pbehne looks like you've got a diff that doesn't run on PRs: https://civet.inl.gov/job/1976337/

pbehne commented:
> @pbehne looks like you've got a diff that doesn't run on PRs: https://civet.inl.gov/job/1976337/

I pushed the fix to my branch. Should work now.

loganharbour commented:
> > @pbehne looks like you've got a diff that doesn't run on PRs: https://civet.inl.gov/job/1976337/
> 
> I pushed the fix to my branch. Should work now.

Where's the PR?
